Refactor IPCache to remove static package-level globals

joestringer opened the pull request:
Meta: https://github.com/cilium/cilium/issues/21142

This PR refactors the IPCache to encapsulate the current package-level global variables into an individual instance of the IPCache, so that subsequent changes can be more easily made to extend the IPCache to hold more information and to resolve conflicts between information more easily. Initialization is moved away from an `init()` function and directly into the daemon.

Structurally this PR first moves the initialization into the daemon, then progressively changes individual packages to rely on that initialized copy of the IPCache, moves other subsystem interactions into the IPCache initialization functions, and finally removes the globals and plumbs the initialized copy of the IPCache from the daemon through to each individual package that needs it.

Fixes: #17849 
Related: #18301

Tasks:
- [x] Rebase
- [x] Fix up `RemoveLabelsExcluded()` so the callers don't have to pass policy, datapath triggerers
- [x] Move `identityMetadata` (and locking friends) into `IPCache`, possibly via dedicated `identityMetadata` structure
- [x] Better documentation
- [x] NewIPCache(config)

joestringer committed:
ipcache: Initialize global later during daemon

Avoid initializing the IPCache statically once in the package, and
instead initialize it as part of the regular daemon startup. This allows
us to inject references to other subsystems as part of the IPCache
initialization, which will lead to code simplification elsewhere in
propagating these references as part of the upcoming patches.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Rely on init for subsystem interactions

The previous commit reworked the ipcache initialization to inject the
policy, datapath subsystem references into the ipcache. This commit
shifts away from passing these same references directly with the calls
to ipcache from various subsystems, instead to rely on the init-time
references to those subsystems.

This cleanup will allow those other subsystems to reduce their reliance
on the implementation details of the ipcache. This patch is not intended
to have any functional impact on the operation of the daemon.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
proxy: Remove references to global ipcache

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
datapath: Remove references to global ipcache

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
endpoint: Remove references to global ipcache

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer committed:
k8s/watchers: Remove references to global ipcache

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
wireguard: Remove references to global ipcache

This one is a bit tricky, because wireguard is initialized before the
datapath, and the datapath before the Daemon. But IPcache is now
initialized much later as part of the Daemon initialization, after
endpoint manager, identity allocator and policy are initialized.

So, get around this by initializing the ipcache as part of the
WireguardAgent.Init() function later on.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Rely on init for identity allocator

Co-authored-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
node/manager: Remove references to global ipcache

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
clustermesh: Remove references to global ipcache

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
daemon: Remove remaining global IPCache references

Signed-off-by: Joe Stringer <joe@cilium.io>

christarazi commented on pkg/ipcache/metadata_test.go:
```diff
@@ -90,18 +90,21 @@ func TestRemoveLabelsFromIPs(t *testing.T) {
+	IPIdentityCache = NewIPCache().
+		WithIdentityAllocator(allocator).
+		WithPolicyHandler(&mockUpdater{}).
+		WithDatapathHandler(&mockTriggerer{})
```
:amazingman:

christarazi reviewed (approved):
:rocket:

christarazi committed:
daemon, ipcache, k8s, node: Embed ipcache metadata into ipcache

This commit furthers the refactor of moving the ipcache away from being
a global object, by embedding the ipcache metadata (IDMD) into the main
ipcache object itself.

Signed-off-by: Chris Tarazi <chris@isovalent.com>

joestringer reviewed (commented):
Latest changes LGTM.

kaworu reviewed (approved):
Hubble changes LGTM.

ldelossa commented on pkg/ipcache/types/types.go:
```diff
@@ -0,0 +1,19 @@
+	"github.com/cilium/cilium/pkg/identity/cache"
+)
+
+type PolicyHandler interface {
```
Can we get some comments for both these interfaces? Especially around any rules or assumptions implementations may need to be aware of.

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -474,9 +474,15 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	// TODO: convert these package level variables to types for easier unit
 	// testing in the future.
-	ipcache.IdentityAllocator = d.identityAllocator
+	ipcache.IPIdentityCache = ipcache.NewIPCache().
```
I generally like this API, but it also leaves these methods available for external usages **outside** of construction. Is that desired here? It may be. 

However, if these methods are **only** meant for construction, then the functional options pattern maybe nicer, where those methods would be typed "options" functions instead of generally available structure methods.

joestringer replied:
It's not desired outside construction.

One of the problems I'm mitigating here is to avoid having to refactor all of the random tests that instantiate an ipcache (but don't really use it). Each time we change the function signatures it's a bit of a pain to go through and update them all. I guess we could equally solve this by passing a `struct ipcacheConfig {...}`[^1] structure as a single parameter, then go through the fields inside to fetch the variables that are used in these builder pattern functions. The caller technically could still forget / avoid initializing that structure though.

In the end, there's basically only one ipcache initialization point that actually matters, and that's the new one in the daemon which does this correctly. So while the interface is open to abuse, in practice we're just not likely to face any maintenance headaches from this approach. :shrug: 

[^1]: ```
    ipcacheConfig { 
        identityAllocator cache.IdentityAllocator                               
        policyHandler     ipcacheTypes.PolicyHandler                            
        datapathHandler   ipcacheTypes.DatapathHandler 
    }
```

joestringer replied:
The more I look at it, the more I think that the ipcacheConfig{...} approach at least gives less of an impression that the fields are optional. With the structure there we can also document that the fields are all required, so then there's a required config option with documentation that says that each of the fields is required (+ semantics / how they fit into ipcache). So maybe I'll look into that.

ldelossa replied:
Sounds good, I didnt know there was a particular scenario you were trying to mitigate against. If that's the case I think that's good enough reason to take this approach, long with your comments about daemon being the only place which inits it.

ldelossa commented on pkg/proxy/proxy.go:
```diff
@@ -96,15 +97,20 @@ type Proxy struct {
+
+	// IPCache is used for tracking IP->Identity mappings and propagating
+	// them to the proxy via NPHDS in the cases described
+	ipcache *ipcache.IPCache
```
nit: this var is shadowing the package name. Maybe a quick rename?

joestringer replied:
This is embedded into the structure, so no shadowing here. I couldn't think of any reasonable renames either, unless we want to get into more obtuse naming like `ipc` or something.

ldelossa commented on daemon/cmd/daemon.go:
```diff
@@ -667,7 +667,7 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	if option.Config.EnableL7Proxy {
 		d.l7Proxy = proxy.StartProxySupport(10000, 20000, option.Config.RunDir,
-			&d, option.Config.AgentLabels, d.datapath, d.endpointManager)
+			&d, option.Config.AgentLabels, d.datapath, d.endpointManager, ipcache.IPIdentityCache)
```
Are we using a global var here? `ipcache.IPIdentityCache` ? Can be OK just a little confused since commit seems to want to remove global usages.

joestringer replied:
Overall the PR first removes global usages from individual packages under `pkg/` commit by commit, which pushes the reference up to the daemon. That way the individual commits can be constrained mostly to only changing a couple of files focused on one area of the codebase. Then later on, one of the patches goes through all of these references in the daemon package only and makes sure they're all referring to the canonical one that the daemon initializes. It looks like you caught this while stepping through commit-by-commit, as the version on the tip of the tree now refers to `d.ipcache` (and also `ipcache.IPIdentityCache` variable is removed, so the build would fail if it was still there at the end of the whole PR).

ldelossa reviewed (approved):
Refactor looks great. 

I'm a little wary about using the structures method set for that "flow like api" for construction *if* those methods are not desired to be called by external code during runtime and in such a way where we return the pointer back to the caller.

jrajahalme commented on pkg/proxy/proxy.go:
```diff
@@ -96,15 +97,20 @@ type Proxy struct {
+
+	// IPCache is used for tracking IP->Identity mappings and propagating
+	// them to the proxy via NPHDS in the cases described
+	ipcache *ipcache.IPCache
```
I don't see this new `ipcache` member being used in this package (maybe it will be used in later commits?).

joestringer replied:
Correct, I moved it to NPHDS and never got rid of this one. Removed, good spotting.

jrajahalme commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -114,6 +117,7 @@ func NewClusterMesh(c Configuration) (*ClusterMesh, error) {
 			Name:      "remote_clusters",
 			Help:      "The total number of remote clusters meshed with the local cluster",
 		}, []string{metrics.LabelSourceCluster, metrics.LabelSourceNodeName}),
+		IPCache: ipcache.NewIPCache(),
```
Creating a new ipcache here is a bit surprising, should this not be passed in as a parameter to `NewClusterMesh()` instead?

joestringer replied:
Ah you're quite right, I got mixed up thinking that this was the clustermesh-apiserver binary. :pray:

jrajahalme commented on pkg/clustermesh/clustermesh.go:
```diff
@@ -91,6 +92,8 @@ type ClusterMesh struct {
 	controllers   *controller.Manager
 	configWatcher *configDirectoryWatcher
 
+	IPCache *ipcache.IPCache
```
Does this need to be exported?

joestringer replied:
Nope, but I'll add a public one into the Configuration object so that this can be configured at init time. Then I'll hide this one.

jrajahalme commented on daemon/cmd/policy.go:
```diff
@@ -465,14 +466,14 @@ func (r *PolicyReactionEvent) Handle(res chan interface{}) {
 // * wait for the regenerations to be finished
 // * upsert or delete CIDR identities to the ipcache, as needed.
-func reactToRuleUpdates(epsToBumpRevision, epsToRegen *policy.EndpointSet, rev uint64, upsertIdentities map[string]*identity.Identity, releasePrefixes []*net.IPNet) {
+func (r *PolicyReactionEvent) reactToRuleUpdates(epsToBumpRevision, epsToRegen *policy.EndpointSet, rev uint64, upsertIdentities map[string]*identity.Identity, releasePrefixes []*net.IPNet) {
```
I wonder if it would be simpler to just make the `Daemon` the receiver here instead of adding pointer to it into `PolicyReactionEvent`? Same for `getIP()`, and maybe others I've not seen yet.

joestringer replied:
yeah, this is a bit simpler, adopted thanks. For getIP, it's just bad abstraction in the ipcache of the `DumpToListenerLocked()` interface. I briefly looked into fixing that ugliness, made a small patch on top, and deferred a bigger refactor for now.

jrajahalme commented on pkg/ipcache/metadata.go:
```diff
@@ -312,8 +311,6 @@ func (ipc *IPCache) RemoveLabelsExcluded(
 	toExclude map[string]struct{},
 	src source.Source,
-	updater ipcacheTypes.PolicyHandler,
-	triggerer ipcacheTypes.DatapathHandler,
```
Minor nit: Looks like these removed parameters were ultimately not used already, so maybe this commit could be squashed with the one that actually removed the use of these parameters in `removeLabelsFromIPs()`?

jrajahalme reviewed (changes requested):
Mostly minor nits, but `pkg/clustermesh` creating it's own ipcache looks like a bug? Maybe grep for calls to `NewIPCache()` to make sure this is only called from the Daemon and test code?

gandro reviewed (approved):
Amazing cleanup. So happy to see this! Nothing to add besides what other reviewers already pointed out. 

As a CODEOWNER, I especially took care review the WireGuard changes, and this looks good to me!

joestringer committed:
proxy: Refactor defaultEndpointInfoRegistry

This implementation was recently extended to make use of ipcache,
refactor it so that it avoids global statically initialized references
so that in later patches we can finally get rid of ipcache globals.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Add DumpToListener

This version hides the lock inside the interface. In future we should
also hide DumpToListenerLocked(), but that probably requires more
invasive refactoring of some of the daemon code to shift it into the
ipcache package.

Signed-off-by: Joe Stringer <joe@cilium.io>

joestringer committed:
ipcache: Refactor initializer for Config parameter

Louis pointed out that the builder pattern being used for ipcache before
was not really appropriate, as the IPCache is not expected to be
modified after initialization, so it's better to pass all of the
parameters up front. Refactor it to take a Configuration object instead
so that the test code can continue to just ignore this parameter where
it's not necessary, while the main initializer will correctly configure
everything that it needs without using unnecessary additional function
calls as part of initialization.

Suggested-by: Louis DeLosSantos <louis@isovalent.com>
Signed-off-by: Joe Stringer <joe@cilium.io>

sayboras commented on pkg/proxy/proxy.go:
```diff
@@ -96,16 +97,26 @@ type Proxy struct {
+
+	// defaultEndpointInfoRegistry is the default instance implementing the
+	// EndpointInfoRegistry interface.
+	defaultEndpointInfoRegistry *endpointInfoRegistry
```
nit: defaultEndpointInfoRegistry -> endpointInfoRegistry

joestringer replied:
I left the default part in there since it seemed like @jrajahalme had some specific intention around that. Probably related to it being only used for the DNS proxy but not for Envoy. :shrug:

sayboras commented on pkg/ipcache/types/types.go:
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package ipcachetypes
```
nit: normally the package name here is same as directory name e.g. types, and import alias will help to avoid any confusion (e.g. ipcacheTypes) like what you have in this PR.

joestringer replied:
I'll add a patch on top for this, the package name was already defined in another file so I can't just fix it up in this one file in the commit that adds this since they all need to be consistent :+1:

sayboras reviewed (approved):
LGTM :100:

joestringer committed:
ipcache/types: Fix package name

Tam points out that the package name should typically follow the name of
the parent directory, but this package wasn't following the ettiquette.
Fix it up.

Signed-off-by: Joe Stringer <joe@cilium.io>
